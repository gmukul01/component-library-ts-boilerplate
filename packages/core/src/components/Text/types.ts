import { HTMLProps, WithThemeProp } from '@core-utils/types';
import { FontSizeTheme, FontWeightTheme } from '@react-basis/theme';

interface Common extends HTMLProps<HTMLElement>, WithThemeProp {
    textColor?: string;
    uppercase?: boolean;
    lineThrough?: boolean;
}

export interface Props extends Common {
    textSize?: keyof FontSizeTheme;
    textWeight?: keyof FontWeightTheme;
}

export interface StyledProps extends Common {
    textSize?: string;
    textWeight?: string;
}
