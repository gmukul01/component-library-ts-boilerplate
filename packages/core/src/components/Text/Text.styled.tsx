import { css, styled } from '@core-styled';
import { defaultTheme } from '@react-basis/theme';
import { Props, StyledProps } from './types';

const uppercase = () => css`
    text-transform: uppercase;
`;

const lineThrough = () => css`
    text-decoration: line-through;
`;

const getMappedProps = ({ theme: { text, font }, ...props }: Props) => {
    const { defaults } = text;
    const { textSize, textWeight, textColor } = props;
    return {
        textSize: font.sizes[textSize || defaults.textSize],
        textWeight: font.weights[textWeight || defaults.textWeight],
        textColor: textColor || 'inherit'
    };
};

export const TextStyled = styled('span').attrs(getMappedProps)<StyledProps>`
    font-size: ${({ textSize }) => textSize};
    font-weight: ${({ textWeight }) => textWeight};
    color: ${({ textColor }) => textColor};

    ${props => props.uppercase && uppercase()};
    ${props => props.lineThrough && lineThrough()};
`;

TextStyled.defaultProps = {
    theme: defaultTheme
};
