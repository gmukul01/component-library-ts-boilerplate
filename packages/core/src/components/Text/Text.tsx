import { WithStyle } from '@core-utils/types';
import React from 'react';
import { TextStyled } from './Text.styled';
import { Props } from './types';

const Text: React.FunctionComponent<Props> & WithStyle = React.forwardRef((props, ref) => {
    const componentType = props.textWeight === 'Strong' ? 'strong' : 'span';
    return (
        <TextStyled {...props} as={componentType} ref={ref}>
            {props.children}
        </TextStyled>
    );
});

Text.displayName = 'Text';
Text.defaultProps = {
    lineThrough: false,
    uppercase: false
};
Text.Style = TextStyled;

export default Text;
