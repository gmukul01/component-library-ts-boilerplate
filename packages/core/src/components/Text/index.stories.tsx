import { FontSizeTheme, FontWeightTheme } from '@react-basis/theme';
import { wInfo } from '@storybook-utils';
import { boolean, color, select } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import Text from './Text';

const sizeOptions: Array<keyof FontSizeTheme> = ['S1', 'M1', 'M2', 'M3', 'L1', 'L2', 'L3', 'L4'];
const weightOptions: Array<keyof FontWeightTheme> = ['Light', 'Normal', 'Strong'];

storiesOf('Components', module).add(
    'Text',
    () => (
        <Text
            uppercase={boolean('Upper Case', false)}
            lineThrough={boolean('Line through', false)}
            textColor={color('Color', 'black')}
            textSize={select<keyof FontSizeTheme>('Text Size', sizeOptions, 'L1')}
            textWeight={select<keyof FontWeightTheme>('Text Weight', weightOptions, 'Normal')}
        >
            Lorem ipsum dolor sit, amet consectetur adipisicing elit. Quaerat excepturi quos vitae quisquam hic molestias et cumque nostrum
            corrupti ad dolorem ab laborum iste veniam, animi optio ipsa quod error!
        </Text>
    ),
    wInfo('Text component with below mentioned props')
);
