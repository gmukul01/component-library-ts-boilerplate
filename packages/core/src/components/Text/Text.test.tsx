import { render } from '@core-test-utils';
import 'jest-styled-components';
import React from 'react';

import Text from './Text';

describe('Text Component', () => {
    it('should render span element by default', () => {
        const { container } = render(<Text />);
        expect(container).toMatchSnapshot();
    });

    it('should render strong element for strong', () => {
        const { container } = render(<Text textWeight="Strong" />);
        expect(container).toMatchSnapshot();
    });

    it('should render with all the props', () => {
        const { container } = render(<Text textSize="M2" textWeight="Light" textColor="pink" uppercase lineThrough />);
        expect(container).toMatchSnapshot();
    });
});
