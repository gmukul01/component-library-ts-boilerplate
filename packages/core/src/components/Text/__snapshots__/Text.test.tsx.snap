// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component should render span element by default 1`] = `
.c0 {
  font-size: 14px;
  font-weight: 400;
  color: inherit;
}

<div>
  <span
    class="c0"
  />
</div>
`;

exports[`Text Component should render strong element for strong 1`] = `
.c0 {
  font-size: 14px;
  font-weight: 700;
  color: inherit;
}

<div>
  <strong
    class="c0"
  />
</div>
`;

exports[`Text Component should render with all the props 1`] = `
.c0 {
  font-size: 14px;
  font-weight: 200;
  color: pink;
  text-transform: uppercase;
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
}

<div>
  <span
    class="c0"
  />
</div>
`;
